### PR TITLE
MSI-2213: Check if product isSalable on MSI standards

### PR DESCRIPTION
**Pull request by sheepfy**
[magento/inventory#2213: Discover pluginization of isSaleable method on product model in Magento](https://github.com/magento/inventory/issues/2213)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/inventory?pullRequest=2604) <br/>All committers have signed the CLA.

File: app/code/Magento/Catalog/Model/Product.php [hunk 1]
```diff
@@ -1729,6 +1729,8 @@ public function setIsDuplicable($value)
      * Check is product available for sale
      *
      * @return bool
+     * @deprecated
```
Comment (ishakhsuvarov) [inline app/code/Magento/Catalog/Model/Product.php]: Unfortunately we would have to deliver this change in the core repository as we cannot accept changes to the core code from Inventory repo. Please submit another PR to [`magento/magento2`](https://github.com/magento/magento2) and link it to this one.

File: app/code/Magento/Catalog/Model/Product.php [hunk 2]
```diff
@@ -1729,6 +1729,8 @@ public function setIsDuplicable($value)
...
+     * @see \Magento\InventorySalesApi\Api\IsProductSalableInterface
```
Comment (ishakhsuvarov) [inline app/code/Magento/Catalog/Model/Product.php]: It might make sense to just link to devdocs to avoid confusing developers having no Inventory Installed.
https://devdocs.magento.com/guides/v2.3/inventory/inventory-api-reference.html

File: app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+use Magento\Framework\Event\Observer;
+use Magento\Framework\Event\ObserverInterface;
+use Magento\InventorySalesApi\Api\Data\SalesChannelInterface;
+use Magento\InventorySalesApi\Api\IsProductSalableInterface;
+use Magento\InventorySalesApi\Api\StockResolverInterface;
+use Magento\Store\Model\StoreManagerInterface;
+
+/**
+ * Class IsProductSalable
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php]: It's best to keep class documentation blocks informative. For example, this observers purpose may be descibed briefly.

File: app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+ *
+ * @package Magento_InventorySales
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php]: `@package` annotation is not used in Magento

File: app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+     */
+    private $storeManager;
+
+    /**
+     * @var array
+     */
+    private $stockIdByWebsite = [];
+
+    /**
+     * IsProductSalable constructor.
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php]: Comment is redundant in class constructors

File: app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        StockResolverInterface $stockResolver,
+        StoreManagerInterface $storeManager
+    ) {
+        $this->isProductSalable = $isProductSalable;
+        $this->stockResolver = $stockResolver;
+        $this->storeManager = $storeManager;
+    }
+
+    /**
+     * Check if product is salable via MSI api
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php]:
```suggestion
     * Customize isSalable method of the Product model to enable it working with multiple stocks
```

File: app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * Retrieve stock_id by website
+     *
+     * @param int $websiteId The website Id
+     * @return int
+     */
+    private function getStockId($websiteId)
+    {
+        if (!isset($this->stockIdByWebsite[$websiteId])) {
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php]: It would be great to avoid introducing in-object cache as it creates technical challenge for anyone attempting to run Magento in multi-threaded environment (for example some sort of solution like [this](https://github.com/appserver-io/appserver)).

What we can do is to introduce a caching facade later on which may be easily disabled with the di configuration.

At this point I would consider this cache to be a slightly premature optimization.

File: app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+            $websiteCode = $this->storeManager->getWebsite($websiteId)->getCode();
+            $stock = $this->stockResolver->execute(SalesChannelInterface::TYPE_WEBSITE, $websiteCode);
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php]: There are extensions out there which provide implementations for different Sales Channels then just Website. It's best to make the type and code parameters customizable from the outside, in this case it may be something like configuring object through the DI.

File: app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+        $salableObject->setIsSalable($isSalable);
+    }
+
+    /**
+     * Retrieve stock_id by website
+     *
+     * @param int $websiteId The website Id
+     * @return int
+     */
+    private function getStockId($websiteId)
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php]: Please use type hinting and include return types where possible.

File: app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+        $salableObject = $observer->getSalable();
+
+        $websiteId = $product->getStore()->getWebsiteId();
+
+        $isSalable = $this->isProductSalable->execute($product->getSku(), $this->getStockId($websiteId));
+        $salableObject->setIsSalable($isSalable);
+    }
+
+    /**
+     * Retrieve stock_id by website
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventorySales/Observer/Product/IsProductSalable.php]:
```suggestion
     * Retrieve Stock ID by Sales Channel
```

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]
It would also be great to complete these changes with a simple integration test

**Comment by ishakhsuvarov** [comment]
Integration test failures should be checked. It looks like some of the older tests used the method in an incorrect way.

**Comment by sidolov** [comment]
@sheepfy please, sign Adobe CLA to proceed with pull request

**Comment by sheepfy** [comment]
> @sheepfy please, sign Adobe CLA to proceed with pull request

done

**Comment by sidolov** [comment]
Hi @sheepfy , I believe it would be better to use plugin for `\Magento\Catalog\Model\Product::isSalable` method instead of observer and don't call the original method at all in case of MSI. Could you please adjust your PR with plugin usage? Thank you!

**Comment by sheepfy** [comment]
yes, i will.

**Comment by sidolov** [comment]
@sheepfy hey, did you have a chance to take a look at it? Thanks!

File: app/code/Magento/InventorySales/Plugin/Catalog/Model/Product.php [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        $this->stockResolver = $stockResolver;
+    }
+
+    /**
+     * @param Subject $subject
+     * @param bool $result
+     */
+    public function afterIsSalable(Subject $subject, $result)
+    {
+        if (!$result) {
```
Comment (sidolov) [inline app/code/Magento/InventorySales/Plugin/Catalog/Model/Product.php]: I believe we should ignore original method behavior since it can be inconsistent with MSI enabled. 
Looks like test failures indicate the issue with it. Please, remove this check and try to use MSI logic only.

**Review by sidolov — CHANGES_REQUESTED** [review]

File: app/code/Magento/InventorySales/Plugin/Catalog/Model/Product.php [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    ) {
+        $this->isProductSalable = $isProductSalable;
+        $this->stockResolver = $stockResolver;
+    }
+
+    /**
+     * @param Subject $subject
+     * @param bool $result
+     */
+    public function afterIsSalable(Subject $subject, $result)
```
Comment (sidolov) [inline app/code/Magento/InventorySales/Plugin/Catalog/Model/Product.php]: Please, add strict types

**Review by sidolov — CHANGES_REQUESTED** [review]

**Comment by sidolov** [comment]
Hi @sheepfy , will you continue progress on the PR?

File: app/code/Magento/InventorySales/Plugin/Catalog/Model/Product.php (hunk 11 above)
Comment (sheepfy) [reply app/code/Magento/InventorySales/Plugin/Catalog/Model/Product.php]: so you want an around plugin ?

Comment (sheepfy) [reply app/code/Magento/InventorySales/Plugin/Catalog/Model/Product.php]: sorry, that my progress is that slow, i was "out of time" in this period

**Comment by sidolov** [comment]
Hi @sheepfy  , due to changes in the repository structure (Magento core code is extracted from the MSI code base) we have changed the base branch to `1.1-develop-legacy`. We will keep this branch for a while to make a smooth transition, in the meanwhile, please, apply changes from the current PR to the new code base structure in `1.1-develop` branch. 

[Installation Guide](https://github.com/magento/inventory/wiki/Metapackage-Installation-Guide) may be helpful for local project development. 

Let me know if you need help or have questions!
Thank you for your collaboration!

**Comment by sidolov** [comment]
Hi @sheepfy , will you continue progress on PR? Thanks!

**Comment by sidolov** [comment]
@sheepfy  I'm closing this pull request since it's targeting legacy branch and cannot be delivered without additional changes because the repository structure has changed. Please, feel free to open the new pull request with the same changes if you wish to continue! Thank you for your collaboration!
